idseq_dag/steps/run_srst2.py

```diff
@@ -69,6 +72,76 @@ def get_common_params(self):
+            grep_params.extend(input_filenames) 
+            grep_output = command.execute_with_output(" ".join(grep_params))
+            output_lines = [line for line in grep_output.split("\n") if line != '']
+            if ":" in output_lines[0]: # for paired fastas
+                read_counts = map(lambda line: int(line.split(":")[1]), output_lines)
+                return reduce(lambda x, y: x + y, list(read_counts))
+            else:
+                return int(output_lines[0])
+        else:
+            wc_params = ['wc', '-l']
+            wc_params.extend(input_filenames)
+            wc_output = command.execute_with_output(" ".join(wc_params))
+            # take the set of characters from the last line, which is the total number of lines
+            # for paired reads or the only line for unpaired reads
+            wc_lines = [line for line in wc_output.split("\n") if line != '']
+            wc_target_line = [line for line in wc_lines[-1].split(" ") if line != '']
+            total_line_count = ''.join(filter(lambda x: x.isdigit(), str(wc_target_line[0])))
+            return int(total_line_count) / 4 # fastqs have 4 lines for every read
+
+    @staticmethod
+    def _append_dpm_to_results(amr_results, total_reads):
+        """Calculates the depth per million for each gene in the result and appends it to the 
+            results dataframe."""
+        amr_results["dpm"] = amr_results.apply(lambda row: row["depth"] * 1000000 / total_reads, axis=1)
+        return amr_results
+
+    @staticmethod
+    def _append_rpm_to_results(proc_amr_results, matched_reads_path, total_reads):
+        """Reads in a table of matched reads generated by generate_mapped_reads_table() and appends
+            the reads per million (rpm) and total reads to the dataframe for processed amr results"""
+        matched_reads = pd.read_csv(matched_reads_path, delimiter="\t", names=["allele", "reads"], usecols=[0,3])
+        matched_reads["allele"] = matched_reads.apply(lambda row: "_".join(row["allele"].split("__")[2:]), axis=1)
+        rpm_list, total_reads_list = PipelineStepRunSRST2._calculate_rpms(matched_reads, proc_amr_results, total_reads)
+        proc_amr_results["total_reads"] = total_reads_list
+        proc_amr_results["rpm"] = rpm_list
+        return proc_amr_results
+
+    @staticmethod
+    def _calculate_rpms(rpm_df, amr_df, total_reads):
+        """Matches each gene in the amr_results dataframe to it's associated number of reads in
+            the matched reads table and calculates the reads per million. Both total reads and
+            reads per million are appended in order in lists and then bulk appended back in
+            _append_rpm_to_results() to the proc_amr_results dataframe."""
+        rpm_list = []
+        total_reads_list = []
+        for row in amr_df.itertuples():
+            reads_for_allele = rpm_df[rpm_df["allele"] == row.allele]["reads"].values[0]
```

Comment:
Are you always guaranteed to find a matching column in the rpm_df here? Worried this line will crash if not.

Reply:
Yes, since rpm_df and amr_df are generated from the same output file they will have the same alleles present.